<a href="https://colab.research.google.com/github/MSagri05/Assignment1-Exploratory-Data-Analysis/blob/main/IAT461_Assignment01_Scaffold.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

---

# ASSIGNMENT 1 BY MANMEET SAGRI
STUDENT # 301545311

# Task 1: Familiarize Yourself with the Dataset

There are two main subtasks:  
1. Load the four data files that are part of this dataset.  
2. Explore each dataset and define the data types of each column. Describe the contents of each column, including the nature of the data values. Follow a similar approach as demonstrated in the lecture and tutorial. Discuss the shape of the data, value ranges, and possible categories. Convert any data into a more appropriate format if needed.  

Add as many code and markdown cells as necessary below. Typically, you should perform one type of investigation per code cell with output, and then comment on the results in a markdown cell below it.  

After completing the analysis, write a concise summary in the **Task 1: summary report** cell at the end. The final report should succinctly summarize the main characteristics of each data file, how they relate to each other, and any notable features.


In [23]:
# IMPORT PANDAS + LOAD DATASETS
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

# LOAD DATASETS
sessions = pd.read_csv("sessions.csv")
viewers = pd.read_csv("viewers.csv")
streamers = pd.read_csv("streamers.csv")
streams = pd.read_csv("streams.csv")

print("Loaded successfully!")

Loaded successfully!


**Sessions shape and first look:**


In [24]:
# INSPECT SESSIONS
# Check how many rows and columns sessions has, and preview the first few rows
print(sessions.shape)
sessions.head()

(22825, 11)


,session_id,viewer_id,streamer_id,stream_id,started_at,ended_at,duration_mins,chat_messages_sent,bits_cheered,followed_during,subscribed_during
0,SES000001,VWR00938,STR0055,SM00860,2024-01-18 19:16:00,2024-01-18 20:28:00,72,0,0,False,False
1,SES000002,VWR00804,STR0034,SM00542,2024-01-04 17:11:00,2024-01-04 19:41:00,150,11,0,False,False
2,SES000003,VWR00704,STR0008,SM00124,2024-02-15 15:11:00,2024-02-15 16:53:00,102,5,0,False,False
3,SES000004,VWR00926,STR0024,SM00373,2024-01-18 19:59:00,2024-01-18 20:56:00,57,3,9663,False,False
4,SES000005,VWR00883,STR0008,SM00120,2024-03-19 20:08:00,2024-03-19 20:22:00,14,7,0,False,False


The sessions dataset has 22,825 rows and 11 columns. Each row is one viewing session, linking a viewer to a streamer and stream via ID columns. It tracks when the session happened, how long it lasted, chat activity, bits cheered, and whether the viewer followed or subscribed during the session.

**Sessions data types and missing values:**


In [25]:
# Check what data type each column is stored as, and how many non-null values exist
sessions.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 22825 entries, 0 to 22824
Data columns (total 11 columns):
 #   Column              Non-Null Count  Dtype 
---  ------              --------------  ----- 
 0   session_id          22825 non-null  object
 1   viewer_id           22825 non-null  object
 2   streamer_id         22765 non-null  object
 3   stream_id           22825 non-null  object
 4   started_at          22825 non-null  object
 5   ended_at            22825 non-null  object
 6   duration_mins       22825 non-null  int64 
 7   chat_messages_sent  22825 non-null  int64 
 8   bits_cheered        22825 non-null  int64 
 9   followed_during     22825 non-null  bool  
 10  subscribed_during   22825 non-null  bool  
dtypes: bool(2), int64(3), object(6)
memory usage: 1.6+ MB


Most columns are complete with no missing values except streamer_id which has 60 missing entries. The timestamp columns are stored as object and will need to be converted to datetime later.

**Sessions numerical summary:**


In [26]:
# Get min, max, mean, median for all numerical columns to understand value ranges
sessions.describe()

,duration_mins,chat_messages_sent,bits_cheered
count,22825.000000,22825.000000,22825.000000
mean,59.051303,6.442629,398.425323
std,60.383979,46.576309,1571.528658
min,0.000000,0.000000,0.000000
25%,15.000000,0.000000,0.000000
50%,39.000000,1.000000,0.000000
75%,83.000000,5.000000,0.000000
max,443.000000,1987.000000,10000.000000


The average session is about 59 minutes but the median is only 39, meaning some long sessions are pulling the average up. Sessions range from 0 to 443 minutes. Most viewers send very few chat messages (median of 1) and most cheer 0 bits, but the max values of 1987 messages and 10000 bits are worth investigating later as potential outliers.

**Convert timestamps in sessions:**

In [27]:
# started_at and ended_at are stored as text (object) — convert them to proper datetime format
sessions["started_at"] = pd.to_datetime(sessions["started_at"])
sessions["ended_at"] = pd.to_datetime(sessions["ended_at"])
sessions.dtypes

,0
session_id,object
viewer_id,object
streamer_id,object
stream_id,object
started_at,datetime64[ns]
ended_at,datetime64[ns]
duration_mins,int64
chat_messages_sent,int64
bits_cheered,int64
followed_during,bool


**Viewers shape and first look:**

In [28]:
# Check how many rows and columns viewers has, and preview the first few rows
print(viewers.shape)
viewers.head()

(1000, 6)


,viewer_id,age_group,country,account_age_days,subscription_tier,preferred_category
0,VWR00001,18-24,AR,355,free,IRL
1,VWR00002,25-34,CA,1670,free,Creative
2,VWR00003,35-44,US,1108,free,Just Chatting
3,VWR00004,45+,NaN,822,tier1,FPS
4,VWR00005,25-34,DE,2226,free,Just Chatting


the viewers dataset has 1000 rows and 6 columns. each row is one viewer with info like their age group, country, account age, subscription tier, and preferred category. i can already see a NaN in the country column for row 3 so there are some missing values there.

**Viewers data types and missing values:**

In [29]:
viewers.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 6 columns):
 #   Column              Non-Null Count  Dtype 
---  ------              --------------  ----- 
 0   viewer_id           1000 non-null   object
 1   age_group           1000 non-null   object
 2   country             920 non-null    object
 3   account_age_days    1000 non-null   int64 
 4   subscription_tier   1000 non-null   object
 5   preferred_category  1000 non-null   object
dtypes: int64(1), object(5)
memory usage: 47.0+ KB


most columns are complete. the only one with missing values is country with 80 missing entries (920 non-null out of 1000). account_age_days is the only numerical column, everything else is stored as object which makes sense since they're all categories.

**Viewers numerical summary + category counts:**

In [30]:
# Numerical summary for account_age_days, and value counts for categorical columns
print(viewers["account_age_days"].describe())
print()
print(viewers["age_group"].value_counts())
print()
print(viewers["subscription_tier"].value_counts())
print()
print(viewers["preferred_category"].value_counts())

count    1000.000000
mean     1881.119000
std      1036.144656
min        33.000000
25%       984.500000
50%      1966.500000
75%      2761.250000
max      3645.000000
Name: account_age_days, dtype: float64

age_group
18-24       334
25-34       289
35-44       134
45+         111
13-17        95
18–24        11
18 to 24     11
25–34         6
35 to 44      4
25 to 34      3
35–44         2
Name: count, dtype: int64

subscription_tier
free          578
tier1         243
tier2         100
tier3          49
premium        10
none            7
TIER1           5
free trial      4
tier4           4
Name: count, dtype: int64

preferred_category
Just Chatting    269
FPS              230
MOBA             161
IRL              120
Creative         117
Sports           103
Name: count, dtype: int64


account ages range from 33 to 3645 days with a median of about 1966 days (~5 years). the 18-24 age group is the largest with 334 viewers. i also noticed the age_group column has inconsistent formatting where the same group appears as "18-24", "18–24", and "18 to 24" which is a data quality issue to fix later. subscription_tier also has messy values like "TIER1", "none", "free trial", and "tier4" which don't belong. most viewers are on free tier (578). Just Chatting is the most preferred category followed by FPS and MOBA.

**Streamers shape and first look:**

In [31]:
# Check how many rows and columns streamers has, and preview the first few rows
print(streamers.shape)
streamers.head()

(76, 7)


,streamer_id,streamer_name,category,language,partner_status,total_followers,avg_concurrent_viewers
0,STR0001,xQc,Just Chatting,English,True,11500000,138706
1,STR0002,Pokimane,Just Chatting,English,True,-25551,71660
2,STR0003,HasanAbi,Just Chatting,English,True,2600000,34220
3,STR0004,Amouranth,Just Chatting,English,True,6200000,68607
4,STR0005,Mizkif,Just Chatting,English,True,2100000,6771


the streamers dataset has 76 rows and 7 columns. each row is one streamer with their name, category, language, partner status, total followers, and average concurrent viewers. i can already see that Pokimane has a negative follower count (-25551) which is clearly wrong and will need to be addressed later.

**Streamers data types and missing values:**

In [32]:
# Check column types and spot any missing values
streamers.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 76 entries, 0 to 75
Data columns (total 7 columns):
 #   Column                  Non-Null Count  Dtype 
---  ------                  --------------  ----- 
 0   streamer_id             76 non-null     object
 1   streamer_name           76 non-null     object
 2   category                76 non-null     object
 3   language                68 non-null     object
 4   partner_status          76 non-null     bool  
 5   total_followers         76 non-null     int64 
 6   avg_concurrent_viewers  76 non-null     int64 
dtypes: bool(1), int64(2), object(4)
memory usage: 3.8+ KB


most columns are complete with no missing values except language which has 8 missing entries (68 non-null out of 76). total_followers and avg_concurrent_viewers are stored as int64 which is correct, and partner_status is boolean which also makes sense.

**Streamers numerical summary:**

In [33]:
# Get value ranges for followers and avg concurrent viewers
streamers.describe()

,total_followers,avg_concurrent_viewers
count,7.600000e+01,76.000000
mean,3.300886e+06,31227.986842
std,3.793772e+06,38307.223652
min,-4.696400e+04,634.000000
25%,7.800000e+05,6633.250000
50%,2.000000e+06,14762.500000
75%,3.875000e+06,36840.750000
max,1.870000e+07,156401.000000


the minimum value for total_followers is -46,964 which is impossible since follower counts can't be negative. this confirms the issue spotted in the previous cell. the max is 18.7 million which shows a huge range between streamers. avg_concurrent_viewers looks more reasonable ranging from 634 to 156,401.

**Streams shape and first look:**

In [34]:
# Check how many rows and columns streams has, and preview the first few rows
print(streams.shape)
streams.head()

(1199, 9)


,stream_id,streamer_id,started_at,ended_at,stream_duration_hrs,category,peak_viewers,title_has_hype_word,was_raid
0,SM00001,STR0001,2024-02-14 18:00:00,2024-02-14 21:00:00,3.0,Just Chatting,108989.0,False,True
1,SM00002,STR0001,2024-01-28 18:00:00,2024-01-28 21:00:00,3.0,Just Chatting,51635.0,True,False
2,SM00003,STR0001,2024-01-31 09:00:00,2024-01-31 16:00:00,7.0,Just Chatting,212127.0,False,False
3,SM00004,STR0001,2024-03-03 14:00:00,2024-03-03 21:00:00,7.0,Just Chatting,186745.0,False,True
4,SM00005,STR0001,2024-01-24 18:00:00,2024-01-24 21:00:00,3.0,Just Chatting,87723.0,False,True


the streams dataset has 1199 rows and 9 columns. each row is one individual broadcast with info about which streamer hosted it, when it started and ended, how long it lasted, the category, peak viewers, and whether the title had a hype word or if it was a raid. looks clean at first glance but we'll check for issues in the next cells.

**Streams data types and missing values:**

In [35]:
# Check column types and spot any missing values
streams.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1199 entries, 0 to 1198
Data columns (total 9 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   stream_id            1199 non-null   object 
 1   streamer_id          1199 non-null   object 
 2   started_at           1199 non-null   object 
 3   ended_at             1199 non-null   object 
 4   stream_duration_hrs  1199 non-null   float64
 5   category             1149 non-null   object 
 6   peak_viewers         1119 non-null   float64
 7   title_has_hype_word  1199 non-null   bool   
 8   was_raid             1199 non-null   bool   
dtypes: bool(2), float64(2), object(5)
memory usage: 68.0+ KB


two columns have missing values: category with 50 missing entries and peak_viewers with 80 missing entries. the timestamp columns are again stored as object and will need to be converted to datetime. everything else looks correct.

**Streams numerical summary:**

In [36]:
# Get value ranges for duration and peak viewers — watch for anything suspicious
streams.describe()

,stream_duration_hrs,peak_viewers
count,1199.000000,1119.000000
mean,4.625104,38375.658624
std,7.995028,50377.085406
min,1.500000,-50.000000
25%,2.500000,7598.500000
50%,3.000000,20224.000000
75%,5.000000,47941.000000
max,96.000000,370165.000000


two suspicious values here. peak_viewers has a minimum of -50 which is impossible since viewer counts can't be negative. stream_duration_hrs has a max of 96 hours which means a stream lasting 4 full days, which seems very unlikely. both are worth investigating in task 2.


**Convert timestamps in streams:**

In [37]:
# started_at and ended_at are stored as text — convert to datetime
streams["started_at"] = pd.to_datetime(streams["started_at"])
streams["ended_at"] = pd.to_datetime(streams["ended_at"])
streams.dtypes

,0
stream_id,object
streamer_id,object
started_at,datetime64[ns]
ended_at,datetime64[ns]
stream_duration_hrs,float64
category,object
peak_viewers,float64
title_has_hype_word,bool
was_raid,bool


**How the four tables link together:**

In [38]:
# Check that the shared ID columns actually overlap between tables
print("Sessions viewer_ids in viewers:", sessions["viewer_id"].isin(viewers["viewer_id"]).sum())
print("Sessions stream_ids in streams:", sessions["stream_id"].isin(streams["stream_id"]).sum())
print("Sessions streamer_ids in streamers:", sessions["streamer_id"].isin(streamers["streamer_id"]).sum())
print("Streams streamer_ids in streamers:", streams["streamer_id"].isin(streamers["streamer_id"]).sum())

Sessions viewer_ids in viewers: 22825
Sessions stream_ids in streams: 22790
Sessions streamer_ids in streamers: 22765
Streams streamer_ids in streamers: 1199


### Task 1 Summary Report

the four datasets together form a complete picture of twitch viewing activity. sessions is the main table linking everything together through viewer_id, stream_id, and streamer_id. viewers describes who is watching, streamers describes the content creators, and streams describes the individual broadcasts. during exploration i spotted several data quality issues including negative follower and viewer counts, inconsistent age group and subscription tier labels, missing values across multiple tables, and broken id relationships between sessions and the other tables. all of these will be addressed in task 2.

# Task 2: Data Cleaning, Diagnostics, and Corrections

Identify and resolve issues in the dataset. These can be of different kinds, such as:
- Look for duplicates or implausible values (e.g., users with negative followers, invalid age groups). Thoroughly document your exploration of the data. Decide whether these are outliers or the result of a systemic problem. If the latter, resolve the problem here; if they are outliers, we will address those later.  
- Define at least three heuristics (common-sense rules) to flag unusual entries (e.g., someone watching dozens of streams per day). Discuss whether what you found are errors or artifacts.  

I will repeat the instructions here, but this is expected for each task: each code cell should focus on a specific analysis. You should run the code, and in the markdown below each code cell and its result, provide a detailed interpretation of what you have uncovered in that step. After completing the analyses, summarize the key findings in the indicated summary report cell below.


In [ ]:
# start code for Task 2 here

### Task 2 summary report

# Task 3: Descriptive Statistics and Visualizations

- Create summary statistics and visualizations to understand user characteristics  
- Create summary statistics for streams data overall, and summaries by age group, account tire, or peak viewers  
- Visualize the distribution of streams durations, popular categories,   
- Show category-specific information and compare streamers based on categories  

In [ ]:
#start code for Task 3 here

### Task 3 Summary report
summarize your findings here.

# Task 4: Dealing with Outliers and Missing Values

- Are there any missing values or outliers in the original data? Perform the analysis and report your findings.  
- Are there any outliers or missing values in the derived data? Perform the analysis and report your findings.  
- If there are no outliers or missing values, introduce 3–5% of missing values in a selected column and 1–2% of outliers in another column.  
- Experiment with two strategies for imputing missing values and two strategies for handling outliers. Show how each strategy affects the characteristics of the data.


In [ ]:
#start code for Task 4 here

### Task 4 Summary report

# Task 5: Correlation and Behavioral Patterns

- Compute pairwise correlations between user features and the number of check-ins  
- Visualize the results using a correlation matrix (use a coolwarm color scheme and proper scaling)  
- Try Spearman rank correlation and explain when it is more appropriate than Pearson

In [ ]:
# tast 5 code starts here

### Task 5 Summary report

# Task 6: Compare Behavior Across User Segments

- Segment users into 2–3 groups (e.g., by age bracket, account tire)  
- Compare their viewing behavior over time using boxplots or dot plots  
- Identify any outliers in group behavior and investigate them

In [ ]:
# task 6 code start here

### Task 6 Summary report

# Task 7: Your Own EDA Questions

Formulate and explore two hypotheses of your own:  
- Propose questions based on curiosity (e.g., “Does a user's variety of categories influence morning vs. evening or session duration?”)  
- Use summary statistics and visualizations to answer each question  
- Briefly discuss what additional data would help you answer the questions more effectively


In [ ]:
# task 7 code start here

### Task 7 Summary report

# Task 8: EDA Presentation (Insight Report)

Summarize your findings in a polished format:  
- Create 3–4 slides or a single-page report highlighting your most important insights  
- Include at least 2 charts and interpret them using clear, plain language  
- Apply Tufte’s principles: minimize chartjunk, maximize data-to-ink ratio, and ensure clear labeling  